Sources/NIOHTTP2/Frame Buffers/OutboundFlowControlBuffer.swift

```diff
@@ -210,17 +267,72 @@ private struct StreamFlowControlState: PerStreamData {
         self.dataBuffer = DataBuffer()
     }
 
-    mutating func nextWrite(maxSize: Int) -> DataBuffer.BufferElement {
+    mutating func bufferWrite(_ write: DataBuffer.BufferElement) {
+        self.dataBuffer.bufferWrite(write)
+    }
+
+    mutating func updateWindowSize(newSize: Int) -> WritabilityState {
+        let oldWindowSize = self.currentWindowSize
+        self.currentWindowSize = newSize
+
+        // If we have no marked writes, nothing changed.
+        guard self.dataBuffer.hasMark else {
+            return .unchanged
+        }
+
+        if oldWindowSize <= 0 && self.currentWindowSize > 0 {
+            // Window opened. We can now write.
+            return .changed(newValue: true)
+        }
+
+        if self.currentWindowSize <= 0 && oldWindowSize > 0 {
+            // Window closed. We can now only write if the first write is zero-sized.
+            if self.dataBuffer.nextWriteIsZeroSized {
+                return .unchanged
+            }
+
+            return .changed(newValue: false)
+        }
+
+        return .unchanged
+    }
+
+    mutating func markFlushPoint() {
+        self.dataBuffer.markFlushPoint()
+    }
+
+    /// Removes all pending writes, invalidating this structure as it does so.
+    mutating func evacuatePendingWrites() -> MarkedCircularBuffer<DataBuffer.BufferElement> {
+        return self.dataBuffer.evacuatePendingWrites()
+    }
+
+    func failAllWrites(error: ChannelError) {
+        self.dataBuffer.failAllWrites(error: error)
+    }
+
+    mutating func nextWrite(maxSize: Int) -> (DataBuffer.BufferElement, WritabilityState) {
```

Comment:
IIUC this should only be called if we're marked, worth a comment/assertion to that end?

Reply:
Yeah, seems like a good idea to me.